### ISSUE TITLE
[Question / Issue][Windows] How to use codegen when "chrome.exe" is GPO blocked?

### ISSUE THREAD

**Issue by placaze** [microsoft/playwright#15327]
I'm using Playwright 1.23.0 on Windows 10 with my work computer. People in my company use Firefox daily and "chrome.exe" is blocked by a Group Policy Object (GPO).

With default settings, "codegen" can't work because "chromium" seems needed and operationnal even if we specify "firefox" as the browser to use.

Is there a way to use "codegen" with 100% Firefox only?

___

I tried another thing with `--channel msedge` but something seems blocked by an admin policy too:
```
$ npx playwright codegen https://www.google.fr --channel msedge
object.<anonymous>: Browser closed.
==================== Browser output: ====================
<launching> C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe --disable-field-trial-config --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-back-forward-cache --disable-breakpad --disable-client-side-phishing-detection --disable-component-extensions-with-background-pages --disable-default-apps --disable-dev-shm-usage --disable-extensions --disable-features=ImprovedCookieControls,LazyFrameLoading,GlobalMediaControls,DestroyProfileOnBrowserClose,MediaRouter,DialMediaRouteProvider,AcceptCHFrame,AutoExpandDetailsElement,CertificateTransparencyComponentUpdater,AvoidUnnecessaryBeforeUnloadCheckSync --allow-pre-commit-input --disable-hang-monitor --disable-ipc-flooding-protection --disable-popup-blocking --disable-prompt-on-repost --disable-renderer-backgrounding --disable-sync --force-color-profile=srgb --metrics-recording-only --no-first-run --enable-automation --password-store=basic --use-mock-keychain --no-service-autorun --export-tagged-pdf --no-sandbox --user-data-dir=C:\Users\username\AppData\Local\Temp\playwright_chromiumdev_profile-HAqe1s --remote-debugging-pipe --no-startup-window
<launched> pid=7816
[pid=7816][err]
[pid=7816][err] DevTools remote debugging is disallowed by the system admin.
[pid=7816][err] [7816:14060:0702/112626.796:ERROR:edge_auth_errors.cc(411)] EDGE_IDENTITY: Get Default OS Account failed: Error: Primary Error: kImplicitSignInFailure, Secondary Error: kAccountProviderFetchError, Platform error: 0, Error string:
[pid=7816][err]
=========================== logs ===========================
<launching> C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe --disable-field-trial-config --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-back-forward-cache --disable-breakpad --disable-client-side-phishing-detection --disable-component-extensions-with-background-pages --disable-default-apps --disable-dev-shm-usage --disable-extensions --disable-features=ImprovedCookieControls,LazyFrameLoading,GlobalMediaControls,DestroyProfileOnBrowserClose,MediaRouter,DialMediaRouteProvider,AcceptCHFrame,AutoExpandDetailsElement,CertificateTransparencyComponentUpdater,AvoidUnnecessaryBeforeUnloadCheckSync --allow-pre-commit-input --disable-hang-monitor --disable-ipc-flooding-protection --disable-popup-blocking --disable-prompt-on-repost --disable-renderer-backgrounding --disable-sync --force-color-profile=srgb --metrics-recording-only --no-first-run --enable-automation --password-store=basic --use-mock-keychain --no-service-autorun --export-tagged-pdf --no-sandbox --user-data-dir=C:\Users\username\AppData\Local\Temp\playwright_chromiumdev_profile-HAqe1s --remote-debugging-pipe --no-startup-window
<launched> pid=7816
[pid=7816][err]
[pid=7816][err] DevTools remote debugging is disallowed by the system admin.
[pid=7816][err] [7816:14060:0702/112626.796:ERROR:edge_auth_errors.cc(411)] EDGE_IDENTITY: Get Default OS Account failed: Error: Primary Error: kImplicitSignInFailure, Secondary Error: kAccountProviderFetchError, Platform error: 0, Error string:
[pid=7816][err]
============================================================
    at Object.<anonymous> (C:\depots_git\OUTILS\tests-e2e-playwright\node_modules\@playwright\test\node_modules\playwright-core\cli.js:17:18)
    at Object.<anonymous> (C:\depots_git\OUTILS\tests-e2e-playwright\node_modules\@playwright\test\cli.js:17:18)
```

___

I found a workaround by renaming "chrome.exe" by "chrome_playwright.exe" in "`C:\Users\username\AppData\Local\ms-playwright\chromium-1012\chrome-win\`" and edit [this line](https://github.com/microsoft/playwright/blob/release-1.23/packages/playwright-core/src/server/registry/index.ts#L57) in "`.\node_modules\@playwright\test\node_modules\playwright-core\lib\server\registry\index.js`" to use "chrome_playwright.exe" instead of "chrome.exe".
I don't really like this hack. 🙁 

"codegen" with Chromium works then. I'm still having a weird "SEC_ERROR_UNKNOWN_ISSUER" issue with codegen Firefox after this workaround, but I don't know if it's related or not.
```
$ npx playwright codegen https://www.google.fr --browser ff
: SEC_ERROR_UNKNOWN_ISSUER
=========================== logs ===========================
navigating to "https://www.google.fr/", waiting until "load"
============================================================
    at captureRawStack (C:\depots_git\OUTILS\tests-e2e-playwright\node_modules\@playwright\test\node_modules\playwright-core\lib\utils\stackTrace.js:62:17) {
  name: 'Error'
}
```

**Comment by rwoll** [comment]
Thanks for the report!

> Is there a way to use "codegen" with 100% Firefox only?

No, when you do:

```
npx playwright codegen --browser firefox
```

it launches Playwright's build of FF as the browser, _and_ a Chromium instance for the codegen UI/controls.

> I found a workaround by renaming "chrome.exe" by "chrome_playwright.exe" in "C:\Users\username\AppData\Local\ms-playwright\chromium-1012\chrome-win\" and edit [this line](https://github.com/microsoft/playwright/blob/release-1.23/packages/playwright-core/src/server/registry/index.ts#L57) in ".\node_modules\@playwright\test\node_modules\playwright-core\lib\server\registry\index.js" to use "chrome_playwright.exe" instead of "chrome.exe".
I don't really like this hack. 🙁

This might break other things in Playwright—I'm not sure :/

> "codegen" with Chromium works then. I'm still having a weird "SEC_ERROR_UNKNOWN_ISSUER" issue with codegen Firefox after this workaround, but I don't know if it's related or not.

Pass `--ignore-https-errors` option to code gen to see if this fixes that. If the site you're visiting is using self-signed or development certs, this could be the source of it. (Or if your company has it's own certs that FF is not picking up). However, since your computer has a blocking policy it might be something else.

Unfortunately, if your work policy is blocking Playwright's browsers and tools, there's limited solutions and help we can provide. It'll be difficult for us to know if it's Playwright causing the issue, or the computer's policy. I recommend working with your admin/IT department to unblock the the Playwright binaries. (See https://playwright.dev/docs/browsers#managing-browser-binaries for the location of where they are.) Once that's done, if you encounter any issues with Playwright itself, please re-open or file a new issue. Thanks!
